sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStage.scala

```diff
@@ -32,143 +31,52 @@ import org.apache.spark.sql.execution.ui.SparkListenerSQLAdaptiveExecutionUpdate
 import org.apache.spark.util.ThreadUtils
 
 /**
- * In adaptive execution mode, an execution plan is divided into multiple QueryStages. Each
- * QueryStage is a sub-tree that runs in a single stage. Before executing current stage, we will
- * first submit all its child stages, wait for their completions and collect their statistics.
- * Based on the collected data, we can potentially optimize the execution plan in current stage,
- * change the number of reducer and do other optimizations.
+ * In adaptive execution mode, an execution plan is divided into multiple QueryStages w.r.t. the
+ * exchange as boundary. Each QueryStage is a sub-tree that runs in a single Spark stage. Before
+ * executing current stage, we will first submit all its child stages, wait for their completions
+ * and collect their statistics. Based on the collected data, we can potentially optimize the
+ * execution plan in current stage, change the number of reducer and do other optimizations.
  */
-abstract class QueryStage extends UnaryExecNode {
-
-  var child: SparkPlan
-
-  // Ignore this wrapper for canonicalizing.
-  override def doCanonicalize(): SparkPlan = child.canonicalized
-
-  override def output: Seq[Attribute] = child.output
-
-  override def outputPartitioning: Partitioning = child.outputPartitioning
-
-  override def outputOrdering: Seq[SortOrder] = child.outputOrdering
+abstract class QueryStage extends LeafExecNode {
```

Comment:
I understand the reasons for changing QueryStage from UnaryExecNode to LeafExecNode are:
- Getting rid of `var` in param list.
- The original approach of UnaryExecNode didn't override `child`.

Is that right?

Reply:
because I removed `QueryStageInput`